### DISCUSSION TITLE
How do I write an if statement with multiple conditions?

### DISCUSSION THREAD

**Discussion by giu67890** [casadi/casadi#3766] [Q&A]
Hi all,

I am trying to write an if statement with multiple conditions in Matlab and for the time being I did not manage to do it, so I am asking for your kind help.
The if statement is the following:
`
if (abs(rotm(2, 1) - rotm(1, 2)) < epsilon && abs(rotm(3, 1) - rotm(1, 3)) < epsilon && abs(rotm(3, 2) - rotm(2, 3)) < epsilon)
`
To give you some context: 

- rotm is a 3x3 matrix
- epsilon is a number, e.g. 1e-12

Thank you for your help.

**Comment by giu67890** [comment]
I forgot to specify that the matrix is composed by MX type variables, whilst epsilon is a number.

**Comment by tmmsartor** [comment]
you can use `logic_and` i.e.:

```
from casadi import *
import numpy as np

x = MX.sym("x")

rotm = MX.sym("m",3,3)
epsilon = 1e-3
if_true = cos(x)
if_false = sin(x)

cond = \
 logic_and(fabs(rotm[1, 0] - rotm[0, 1]) < epsilon,
 logic_and(fabs(rotm[2, 0] - rotm[0, 2]) < epsilon,
           fabs(rotm[2, 1] - rotm[1, 2]) < epsilon))

y = if_else(cond,if_true,if_false)
f = Function("f", [x,rotm],[y])
print(f(pi/2, np.ones((3,3))))
nonsym = np.ones((3,3))
nonsym[1,0] = 2
print(f(pi/2, nonsym))
```

If you feel more creative also the following should work:
```
cond = \
  sign(fabs(rotm[1, 0] - rotm[0, 1]) - epsilon)+ \
  sign(fabs(rotm[2, 0] - rotm[0, 2]) - epsilon)*10+ \
  sign(fabs(rotm[2, 1] - rotm[1, 2]) - epsilon)*100+111
y = if_else(cond,if_false,if_true)
```

**Reply by giu67890** [reply]
Hi @tmmsartor,

Firstly, thanks for your answer.
Let me start by saying that I'm not a CasADi expert and that's probably the reason why I'm struggling to get your answer and apply it to my code.
That "if" branch is one of some others present in this function I wrote some days ago and I'm gonna copy and paste it to let you understand (I'm using CasADi with Matlab):

```
function axis_angle_vector = compute_axis_angles_rotm(rotm)   
 
    epsilon = 1e-12;
    epsilon2 = epsilon * 10;

    % Check singularities
    if (abs(rotm(2, 1) - rotm(1, 2)) < epsilon && ...
            abs(rotm(3, 1) - rotm(1, 3)) < epsilon && ...
            abs(rotm(3, 2) - rotm(2, 3)) < epsilon)
        
        % Singularity found

        if (abs(rotm(2, 1) + rotm(1, 2)) < epsilon2 && ...
            abs(rotm(3, 1) + rotm(1, 3)) < epsilon2 && ...
            abs(rotm(3, 2) + rotm(2, 3)) < epsilon2 && ...
            abs(rotm(1, 1) + rotm(2, 2) + rotm(3, 3) - 3) < epsilon2)
            
            % This singularity is identity matrix so angle = 0

            angle = 0;

            axis = [0; 0; 1];

            axis_angle_vector = axis * angle;

        else

            % The singularity is angle = 180°
            
            angle = pi;

            xx = (rotm(1, 1) + 1) / 2;
            yy = (rotm(2, 2) + 1) / 2;
            zz = (rotm(3, 3) + 1) / 2;
            xy = (rotm(2, 1) + rotm(1, 2)) / 4;
            xz = (rotm(3, 1) + rotm(1, 3)) / 4;
            yz = (rotm(3, 2) + rotm(2, 3)) / 4;

            if((xx > yy) && (xx > zz))
            
                % R11 is the largest diagonal term
                x = sqrt(xx);
                y = xy / x;
                z = xz / x;
                
            elseif(yy > zz)
                
                % R22 is the largest diagonal term
                y = sqrt(yy);
                x = xy / y;
                z = yz / y;

            else
                
                % R33 is the largest diagonal term
                z = sqrt(zz);
                x = xz / z;
                y = yz / z;

            end

            axis = [x; y; z];

            axis_angle_vector = axis * angle;

        end

    else
       
        f = (rotm(1, 1) + rotm(2, 2) + rotm(3, 3) - 1) / 2;

        x = rotm(2, 3) - rotm(3, 2);
        y = rotm(3, 1) - rotm(1, 3);
        z = rotm(1, 2) - rotm(2, 1);

        axis = [x; y; z];
        axis_norm = norm(axis);
        angle = atan2(axis_norm / 2, f);

        % Normalizing axis
        if(axis_norm < epsilon)

            axis = [1; 0; 0];

        else
            
            axis = axis / axis_norm;

        end
        
        axis_angle_vector = axis * angle;
    end
end
```

I am calling this function from another part of the code and the 'rotm' variable, as said, is a 3x3 matrix of MX type.
What I don't understand is how to use the _if_else_ function (or others) correctly into my code. For the time being, I am receiving a conversion error because a MX type cannot be converted into a logical one.
Maybe I can enclose the code in the if and else branches into two functions and try to use them, but given that the presence of more than one branches dependent on MX variables, it would be a little messy because I would need to create too many functions. 

Thanks for your help

**Reply by tmmsartor** [reply]
Yes, the solution I was proposing require you to create an expression (no need for functions) for every branch, this require slightly more code and the coding-style will be different. 
Not sure if there are other solution to translate this exact formulation in CasADi.
One thing you may consider is to try another way to represent rotations which is more differentiation friendly i.e. quaternions.

**Reply by giu67890** [reply]
@tmmsartor I have managed to transform the conditional branches so that MX type variables can be used. I think that could be a nice workaround. It consists into transforming the _if_ into boolean functions. Let me explain:

The condition is written in this way:
`cond_is_singularity = (abs(rotm(2, 1) - rotm(1, 2)) < epsilon) * ...
            (abs(rotm(3, 1) - rotm(1, 3)) < epsilon) * ...
            (abs(rotm(3, 2) - rotm(2, 3)) < epsilon);`
            
In this way I can use it just like this:

`axis_angle_vector = ~cond_is_singularity * else_branch_no_singularity(rotm, epsilon) + ...
            cond_is_singularity * [...] + ...`
            
I hope that this solution can be useful to someone in this specific situation.

**Reply by tmmsartor** [reply]
Nice idea and thanks for sharing! It is good to know this is working for you.

**Reply by goncalosousatorres** [reply]
> @tmmsartor I have managed to transform the conditional branches so that MX type variables can be used. I think that could be a nice workaround. It consists into transforming the _if_ into boolean functions. Let me explain:
> 
> The condition is written in this way: `cond_is_singularity = (abs(rotm(2, 1) - rotm(1, 2)) < epsilon) * ... (abs(rotm(3, 1) - rotm(1, 3)) < epsilon) * ... (abs(rotm(3, 2) - rotm(2, 3)) < epsilon);`
> 
> In this way I can use it just like this:
> 
> `axis_angle_vector = ~cond_is_singularity * else_branch_no_singularity(rotm, epsilon) + ... cond_is_singularity * [...] + ...`
> 
> I hope that this solution can be useful to someone in this specific situation.

This was useful, thank you!
